plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryMetadata.java

```diff
@@ -1084,6 +1086,33 @@ public Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(
         return Optional.of(new ConstraintApplicationResult<>(updatedHandle, remainingFilter, constraint.getExpression(), false));
     }
 
+    @Override
+    public Optional<LimitApplicationResult<ConnectorTableHandle>> applyLimit(ConnectorSession session, ConnectorTableHandle handle, long limit)
+    {
+        BigQueryTableHandle table = (BigQueryTableHandle) handle;
+
+        if (table.limit().isPresent() && table.limit().get() <= limit) {
+            return Optional.empty();
+        }
+
+        boolean pushdownLimit = false;
+
+        if (table.isNamedRelation()) {
+            BigQueryNamedRelationHandle namedRelationHandle = table.getRequiredNamedRelation();
+            // Storage API doesn't support limiting rows, but REST API supports LIMIT clause.
+            // So pushdown limit only for rest API.
+            pushdownLimit = !namedRelationHandle.isUseStorageApi();
+        }
+        else if (table.isQueryRelation()) {
+            pushdownLimit = true;
+        }
+        if (!pushdownLimit) {
+            return Optional.empty();
+        }
```

Comment:
Can we simplify them into a single if condition - Ideally we need to push down query for REST API approach - so if `table.isQueryRelation || (table.isNamedRelation() && !table.isUseStorageApi)` - this can be a static method which we could be reused at `SplitSource` WDYT ?